addons/l10n_eg_edi_eta/models/account_edi_format.py

```diff
@@ -0,0 +1,387 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+
+import json
+from base64 import b64encode
+
+
+from odoo import models, api, _
+import logging
+import requests
+from werkzeug.urls import url_quote
+_logger = logging.getLogger(__name__)
+
+
+ETA_DOMAINS = {
+    'preproduction': 'https://api.preprod.invoicing.eta.gov.eg',
+    'production': 'https://api.invoicing.eta.gov.eg',
+    'token.preproduction': 'https://id.preprod.eta.gov.eg',
+    'token.production': 'https://id.eta.gov.eg',
+}
+
+
+
+class AccountEdiFormat(models.Model):
+    _inherit = 'account.edi.format'
+
+    @api.model
+    def _l10n_eg_get_eta_api_domain(self):
+        return self.env.company.l10n_eg_production_env and ETA_DOMAINS['production'] or ETA_DOMAINS['preproduction']
+
+    @api.model
+    def _l10n_eg_get_eta_token_domain(self):
+        return self.env.company.l10n_eg_production_env and ETA_DOMAINS['token.production'] or ETA_DOMAINS['token.preproduction']
+
+    @api.model
+    def _l10n_eg_eta_connect_to_server(self, request_data, request_url, method, api_domain=False):
+        if not api_domain:
+            api_domain = self._l10n_eg_get_eta_api_domain()
+        request_url = request_url % api_domain
```

Comment:
It would be preferable to make sure the hostname or URL cannot be chosen arbitrarily by the caller. Could we remove the placeholder `%s` in the `request_url` param, and force-prepend the hostname here based on the `api_domain`? This way the caller cannot pass an arbitrary URL like `https://somethingelse.com#ignorethis=%s`.

Reply:
this still needs to be fixed